File: 스택/rian00/2504.js
const { sep } = require("path");

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .trim()
  .split("\n");

const [str] = input;

const stack = [];
let flag = false;
let total = 0;
for (const e of str) {
  flag = false;
  switch (e) {
    case "(":
    case "[":
      stack.push(e);
      break;
    case ")":
      if (stack.length > 0 && stack[stack.length - 1] === "(") {
        stack.pop();
      } else {
        console.log(0);
        flag = true;
      }
      break;
    case "]":
      if (stack.length > 0 && stack[stack.length - 1] === "[") {
        stack.pop();
      } else {
        console.log(0);
        flag = true;
      }
      break;
  }
  if (flag) break;
}
if (!flag && stack.length === 0) console.log("YES");
else if (!flag && stack.length > 0) console.log(0);

File: 스택의 활용/JunYoungKr/2504.py
from collections import deque
import sys

input = sys.stdin.readline

str = list(input().strip())
print(str)

# 결과값
result = 0

stack = []

for i in str:
    # 여는 괄호라면 stack에 넣음
    if i in '([':
        stack.append(i)
        continue
    
    # stack의 길이가 0이 아닌 경우
    if len(stack) != 0:
        # stack.pop() 과 i 비교
        if stack[-1] == '(':
            # 짝이 이뤄지는 경우
            if i == ')':
                result += 2
                stack.pop()
            # 같은 괄호인 경우
            elif i == '(':
                stack.append(i)
        elif stack[-1] == '[':
            if i == ']':
                result += 3
                stack.pop()
            elif i == '[':
                stack.append(i)
    print("stack: ", stack)
    print("result: ", result)
        
    
                
        
print(result)          
        


# (()[[]])([])

# 연산의 끝
# stack.pop == '('이고 i == ')' 일 때 result += 2
# stack.pop == '['이고 i == '